File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int s=0;
        int n=nums.size();
        int e=n-1;
        while(s<=e)
        {
            int mid=s+(e-s)/2;
            if(nums[mid]==target)
            {
                return true;
            }
            if(nums[s]==nums[mid] && nums[mid]==nums[e])
            {
                s++;
                e--;
            }
            else if(nums[mid]>=nums[s])
            {
                if(target>=nums[s] && target<nums[mid])
                {
                    e=mid-1;
                }
                else
                {
                    s=mid+1;
                }
            }
            else
            {
                if(target<=nums[e] && target>nums[mid])
                {
                    s=mid+1;
                }
                else
                {
                    e=mid-1;
                }
            }
        }
        return false;
    }
};

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        s=0
        n=len(nums)
        e=n-1
        while(s<=e):           
            mid=s+(e-s)//2
            if(nums[mid]==target):
                return True
            if(nums[s]==nums[mid] and nums[mid]==nums[e]):
                s+=1
                e-=1
            elif(nums[mid]>=nums[s]):
                if(target>=nums[s] and target<nums[mid]):
                    e=mid-1
                else:
                    s=mid+1
            else:
                if(target<=nums[e] and target>nums[mid]):
                    s=mid+1
                else:
                    e=mid-1
        return False